File: python/ml/embed.py
"""embed() — text to features (internal module, not part of public API).

Explicit, stateful text embedding. Returns an Embedder that stores
the fitted tokenizer/vocabulary. Use ml.tokenize() for the public API.

Direct import:
    >>> from ml.embed import embed, Embedder
    >>> emb = embed(texts, method="tfidf")
"""

from __future__ import annotations

from dataclasses import dataclass

import pandas as pd


@dataclass
class Embedder:
    """Fitted text embedder.

    Stores the fitted tokenizer/vectorizer and provides transform() for new texts.

    Attributes
    ----------
    vectors : pd.DataFrame
        Embedded feature vectors from training texts (n_samples x vocab_size)
    method : str
        Embedding method used ("tfidf", "sbert", "bilstm")
    vocab_size : int
        Vocabulary size (number of features)
    _vectorizer : Any
        Internal: fitted sklearn vectorizer or embedding model
    """

    vectors: pd.DataFrame
    method: str
    vocab_size: int
    _vectorizer: object

    def transform(self, texts: pd.Series | list) -> pd.DataFrame:
        """Transform new texts using the fitted embedder.

        Parameters
        ----------
        texts : pd.Series or list
            New texts to embed (uses stored vocabulary/tokenizer)

        Returns
        -------
        pd.DataFrame
            Embedded feature vectors (n_samples x vocab_size)

        Examples
        --------
        >>> import ml
        >>> import pandas as pd
        >>> texts = pd.Series(["good product", "bad service", "great value"])
        >>> emb = ml.embed(texts, method="tfidf")
        >>> new_texts = pd.Series(["excellent quality"])
        >>> new_vectors = emb.transform(new_texts)
        >>> new_vectors.shape
        (1, vocab_size)
        """
        from ._types import DataError

        # Convert list to Series if needed
        if isinstance(texts, list):
            texts = pd.Series(texts)

        # Validate
        if not isinstance(texts, pd.Series):
            raise DataError(f"Expected pd.Series or list, got {type(texts).__name__}")

        # Transform using stored vectorizer
        if self.method == "tfidf":
            embeddings = self._vectorizer.transform(texts)
            # Convert to DataFrame (our TfidfVectorizer returns dense ndarray)
            df = pd.DataFrame(
                embeddings, columns=[f"tfidf_{i}" for i in range(self.vocab_size)]
            )
            return df
        else:
            # Other methods (sbert, bilstm) would go here
            raise NotImplementedError(
                f"method='{self.method}' not yet implemented. Use method='tfidf'."
            )


def embed(
    texts: pd.Series | list,
    *,
    method: str = "tfidf",
    max_features: int = 100,
) -> Embedder:
    """Embed texts into numeric features.

    Explicit, stateful embedding. Returns an Embedder that stores the fitted
    tokenizer/vocabulary for use at prediction time.

    Parameters
    ----------
    texts : pd.Series or list
        Texts to embed
    method : str, default="tfidf"
        Embedding method. Options:
        - "tfidf": TF-IDF (sklearn, fast, default)
        - "sbert": Sentence-BERT (semantic, optional dependency)
        - "bilstm": Learned BiLSTM (optional dependency)
    max_features : int, default=100
        Maximum vocabulary size (number of features to generate)

    Returns
    -------
    Embedder
        Fitted embedder with .vectors (DataFrame), .transform(new_texts),
        .method, .vocab_size attributes.

    Raises
    ------
    DataError
        If texts is not pd.Series or list, or if empty
    ConfigError
        If method is not recognized

    Examples
    --------
    >>> import ml
    >>> import pandas as pd
    >>> texts = pd.Series(["good product", "bad service", "great value"])
    >>> emb = ml.embed(texts, method="tfidf")
    >>> emb.vectors.shape
    (3, 100)
    >>> emb.method
    'tfidf'
    >>> emb.vocab_size
    100

    At prediction time:
    >>> new_texts = pd.Series(["excellent quality"])
    >>> new_vectors = emb.transform(new_texts)
    >>> new_vectors.shape
    (1, 100)

    Save and load:
    >>> ml.save(emb, "embedder.ml")
    >>> emb_loaded = ml.load("embedder.ml")
    """
    from ._types import ConfigError, DataError

    # Convert list to Series if needed
    if isinstance(texts, list):
        texts = pd.Series(texts)

    # Validation
    if not isinstance(texts, pd.Series):
        raise DataError(f"Expected pd.Series or list, got {type(texts).__name__}")

    if len(texts) == 0:
        raise DataError("Cannot embed empty texts (0 samples)")

    # Supported methods (only tfidf in v1.0)
    if method not in ("tfidf",):
        raise ConfigError(
            f"method='{method}' not recognized. Choose from: ['tfidf']"
        )

    # TF-IDF implementation (Gate 1)
    if method == "tfidf":
        from ._transforms import TfidfVectorizer

        vectorizer = TfidfVectorizer(max_features=max_features, lowercase=True)
        embeddings = vectorizer.fit_transform(texts)

        # Convert to DataFrame (our TfidfVectorizer returns dense ndarray)
        actual_vocab_size = embeddings.shape[1]
        vectors_df = pd.DataFrame(
            embeddings,
            columns=[f"tfidf_{i}" for i in range(actual_vocab_size)],
        )

        return Embedder(
            vectors=vectors_df,
            method=method,
            vocab_size=actual_vocab_size,
            _vectorizer=vectorizer,
        )

    else:
        # Should not reach here after validator — future methods deferred
        raise NotImplementedError(f"method='{method}' not yet implemented.")

File: ml.Rcheck/00_pkg_src/ml/R/embed.R
#' Embed texts into numeric features
#'
#' Fits a text vectorizer on training texts and returns an embedder object
#' that stores the vocabulary for consistent transform at prediction time.
#'
#' Currently supports TF-IDF ('tm' package). SBERT and neural methods are
#' planned for future gates.
#'
#' @param texts A character vector of texts to embed
#' @param method Embedding method. Currently only "tfidf" is supported.
#' @param max_features Maximum vocabulary size (number of TF-IDF features).
#'   Default 100.
#' @returns An object of class `ml_embedder` with:
#'   - `$vectors`: data.frame of TF-IDF features (n_texts x max_features)
#'   - `$method`: the method used
#'   - `$vocab_size`: number of features generated
#'   - `$transform(new_texts)`: apply stored vocabulary to new texts
#' @export
#' @examples
#' \donttest{
#' # Requires 'tm' package: install.packages("tm")
#' texts <- c("good product", "bad service", "great value", "poor quality")
#' emb <- ml_embed(texts, method = "tfidf", max_features = 20)
#' emb$vocab_size
#' nrow(emb$vectors)
#'
#' # Transform new texts using the fitted vocabulary
#' new_texts <- c("excellent quality", "terrible service")
#' new_vecs <- emb$transform(new_texts)
#' }
ml_embed <- function(texts, method = "tfidf", max_features = 100L) {
  .embed_impl(texts = texts, method = method, max_features = max_features)
}

.embed_impl <- function(texts, method = "tfidf", max_features = 100L) {
  if (!requireNamespace("tm", quietly = TRUE)) {
    config_error("'tm' package required for embed(). Install with: install.packages('tm')")
  }

  # Validate inputs
  if (is.null(texts) || length(texts) == 0L) {
    data_error("Cannot embed empty texts (0 samples)")
  }
  if (!is.character(texts)) {
    data_error(paste0("Expected character vector, got ", class(texts)[[1]]))
  }
  if (!method %in% c("tfidf")) {
    config_error(paste0(
      "method='", method, "' not recognized. Choose from: tfidf"
    ))
  }

  max_features <- as.integer(max_features)

  if (method == "tfidf") {
    # Build corpus
    corpus <- tm::Corpus(tm::VectorSource(texts))
    corpus <- tm::tm_map(corpus, tm::content_transformer(tolower))
    corpus <- tm::tm_map(corpus, tm::removePunctuation)
    corpus <- tm::tm_map(corpus, tm::removeNumbers)
    corpus <- tm::tm_map(corpus, tm::stripWhitespace)

    # Build DTM with TF-IDF weighting
    dtm <- tm::DocumentTermMatrix(
      corpus,
      control = list(
        weighting    = tm::weightTfIdf,
        bounds       = list(global = c(1L, Inf)),
        wordLengths  = c(2L, Inf)
      )
    )

    # Convert to dense matrix (allows base R operations)
    mat_full   <- as.matrix(dtm)

    # Limit to max_features most frequent terms (by total TF-IDF weight)
    col_sums   <- colSums(mat_full)
    top_idx    <- order(col_sums, decreasing = TRUE)[seq_len(min(max_features, ncol(mat_full)))]
    top_terms  <- colnames(mat_full)[top_idx]
    mat        <- mat_full[, top_idx, drop = FALSE]

    vocab_size <- ncol(mat)
    colnames(mat) <- paste0("tfidf_", seq_len(vocab_size))
    vectors_df <- as.data.frame(mat)

    # Store vocabulary and IDF weights for transform()
    # doc_freq = number of documents containing each term
    doc_freq    <- colSums(mat_full[, top_idx, drop = FALSE] > 0)
    idf_weights <- log((nrow(mat) + 1L) / (doc_freq + 1L)) + 1L  # sklearn-style smooth IDF
    names(idf_weights) <- top_terms
    vocab       <- top_terms

    embedder <- new_ml_embedder(
      vectors    = vectors_df,
      method     = method,
      vocab_size = vocab_size,
      vocab      = vocab,
      idf        = idf_weights
    )
    return(embedder)
  }

  config_error(paste0("method='", method, "' not yet implemented."))
}

# ── S3 type ────────────────────────────────────────────────────────────────────

#' @keywords internal
new_ml_embedder <- function(vectors, method, vocab_size, vocab, idf) {
  # transform closure captures vocab and idf
  transform_fn <- function(new_texts) {
    if (!requireNamespace("tm", quietly = TRUE)) {
      config_error("'tm' required. Install with: install.packages('tm')")
    }
    if (!is.character(new_texts) || length(new_texts) == 0L) {
      data_error("new_texts must be a non-empty character vector")
    }
    corpus <- tm::Corpus(tm::VectorSource(new_texts))
    corpus <- tm::tm_map(corpus, tm::content_transformer(tolower))
    corpus <- tm::tm_map(corpus, tm::removePunctuation)
    corpus <- tm::tm_map(corpus, tm::removeNumbers)
    corpus <- tm::tm_map(corpus, tm::stripWhitespace)

    dtm_new <- tm::DocumentTermMatrix(
      corpus,
      control = list(
        dictionary   = vocab,
        weighting    = tm::weightTf,
        wordLengths  = c(2L, Inf)
      )
    )

    # Apply stored IDF weights manually (TF × IDF)
    mat <- as.matrix(dtm_new)
    # Reorder/fill columns to match stored vocab
    aligned <- matrix(0, nrow = nrow(mat), ncol = length(vocab))
    colnames(aligned) <- vocab
    shared <- intersect(colnames(mat), vocab)
    if (length(shared) > 0L) aligned[, shared] <- mat[, shared, drop = FALSE]

    # Apply IDF
    for (j in seq_along(vocab)) {
      aligned[, j] <- aligned[, j] * idf[[vocab[[j]]]]
    }

    df_out <- as.data.frame(aligned)
    colnames(df_out) <- paste0("tfidf_", seq_len(ncol(df_out)))
    df_out
  }

  structure(
    list(
      vectors    = vectors,
      method     = method,
      vocab_size = vocab_size,
      vocab      = vocab,
      idf        = idf,
      transform  = transform_fn
    ),
    class = "ml_embedder"
  )
}

#' Print ml_embedder
#' @param x An ml_embedder object
#' @param ... Ignored
#' @returns The object \code{x}, invisibly.
#' @export
print.ml_embedder <- function(x, ...) {
  cat(sprintf("-- Embedder [%s] --\n", x[["method"]]))
  cat(sprintf("  texts     : %d\n", nrow(x[["vectors"]])))
  cat(sprintf("  vocab_size: %d\n", x[["vocab_size"]]))
  cat(sprintf("  use $transform(new_texts) to embed new data\n"))
  cat("\n")
  invisible(x)
}
